--- a/src/SALT2mu.c
+++ b/src/SALT2mu.c
@@ -14366,17 +14366,18 @@
     ig = (int)INFO_BIASCOR.IG[i]; 
 
     NTMP = NUSE[ia][ib][ig];
+    if ( NTMP >= NROW_malloc ) {
+      sprintf(c1err,"NUSE[ia,ib,ig=%d,%d,%d] = %d is outside allocated indices",
+	      ia, ib, ig, NTMP );
+      sprintf(c2err,"NROW_malloc = %d", NROW_malloc);
+      errlog(FP_STDOUT, SEV_FATAL, fnam, c1err, c2err);     
+    }
     MUDIF[ia][ib][ig][NTMP]   = muDif ;
     MUERRSQ[ia][ib][ig][NTMP] = muErrsq ;
     WGT_POP[ia][ib][ig][NTMP] = WGT_biasCor_population(i,fnam);   
     NUSE[ia][ib][ig]++ ;
     
-    if ( NUSE[ia][ib][ig] >= NROW_malloc ) {
-      sprintf(c1err,"NUSE[ia,ib,ig=%d,%d,%d] = %d exceeds malloc size",
-	      ia, ib, ig, NUSE[ia][ib][ig] );
-      sprintf(c2err,"NROW_malloc = %d", NROW_malloc);
-      errlog(FP_STDOUT, SEV_FATAL, fnam, c1err, c2err);     
-    }
+
   
   } // end loop over biasCor sample
 
